Fix Dimension impl for Circle

rfuest opened the pull request:
Hi! Thank you for helping out with Embedded Graphics development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a simulator example(s) where applicable
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) and appropriate crate (`embedded-graphics`, `simulator`, `tinytga`, `tinybmp`) if your changes affect the **public API**
- [x] Run `rustfmt` on the project
- [ x Run `./build.sh` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

This PR fixes the `Dimension` impl for `Circle`s. The size of the bounding box was off by one before and was inconsistent with the impl for `Rectangle`.

In addition the PR also reduces the code duplication in the iterators impls by adding an internal `DistanceIterator`.

jamwaffles commented on embedded-graphics/src/primitives/circle.rs:
```diff
@@ -274,11 +244,11 @@ where
 }
 
-fn diameter_to_threshold(diameter: i32) -> i32 {
+fn diameter_to_threshold(diameter: u32) -> i32 {
```
I think this should return a `u32`, i.e. `fn diameter_to_threshold(diameter: u32) -> u32` along with the other `i32` to `u32` fixes in this file to make it compile. The only change beyond type fixes would be that `DistanceIterator` becomes something like:

```diff
impl Iterator for DistanceIterator {
    type Item = (Point, u32);

    fn next(&mut self) -> Option<Self::Item> {
        self.points.next().map(|p| {
            let delta = self.center - p * 2;
-           let distance = delta.x.pow(2) + delta.y.pow(2);
+           let distance = delta.x.pow(2) as u32 + delta.y.pow(2) as u32;

            (p, distance)
        })
    }
}
```

(or maybe that should be `let distance = (delta.x.pow(2) as u32).saturating_add(delta.y.pow(2) as u32);` idk)

This makes it clearer to the reader that thresholds can never be negative, which percolates through the rest of the file.

rfuest replied:
That's a good point. I've changed all thresholds and distances to `u32`.

rfuest commented:
I've addressed your comment and also tweaked the `StyledCircleIterator` a bit. You can review it whenever you have some time and energy left, I don't mind waiting a while.

jamwaffles commented on embedded-graphics/src/primitives/circle.rs:
```diff
@@ -177,91 +163,84 @@ impl Points {
         // If the fill and stroke colors are `None`, treat entire object as transparent and exit
         // early.
-        if self.style.stroke_color.is_none() && self.style.fill_color.is_none() {
+        if self.outer_color.is_none() && self.inner_color.is_none() {
             return None;
         }
```
This might be a bit more pleasant:

```suggestion
        self.outer_color.or(self.inner_color)?;
```

This also optimises to like 2 fewer asm instructions so yay I guess :grinning:

rfuest replied:
That's much nicer. But I've decided to remove it completely, which should remove some additional asm instructions, but i haven't counted :wink:

rfuest commented:
> Interestingly, `cargo bench circle` is roughly 50% faster in this branch compared to `master`. This is on a laptop so results are a bit noisy but it's an interesting positive side effect. YMMV of course.

That's a nice surprise, I had feared that this code might be slower than the old version. But I get the same ~50% improvement on my PC.

jamwaffles commented:
Good stuff!
